File: typescript/webSocket/src/app/share/Api/StateMachine/StateMachine.ts
import {EventAction, EventNode, IState} from './IState';
import {BaseContext} from './BaseContext';

/**
 * A simple state machine.
 */
export class StateMachine<TState, TEvent, TCtx extends BaseContext<TState, TEvent, TParam>, TParam> {

  stateLookup = new Map<TState, StateNode<TState, TEvent, TCtx, TParam>>();

  /**
   * Used to execute the transition.
   * @param ctx The current context.
   * @param event The event
   * @param param The parameter associated with the event.
   */
  transition(
    ctx: TCtx,
    event: TEvent,
    param: TParam) {
    ctx.add(event, param);
    this.handleEvent(ctx);
  }

  /**
   * Used to add a state to the state machine.
   * @param state The state to add to the state machine.
   */
  addState(state: IState<TState, TEvent, TCtx, TParam>): StateMachine<TState, TEvent, TCtx, TParam> {
    if (this.stateLookup.has(state.state())) {
      throw new Error(`State has already been registered ${state.state()}`);
    } else {
      this.stateLookup.set(state.state(), new StateNode<TState, TEvent, TCtx, TParam>(state));
    }
    return this;
  }

  /**
   * used to handle to the next event.
   * @param ctx The context for the state machine.
   */
  private handleEvent(
    ctx: TCtx
  ) {
    const state = this.stateLookup.get(ctx.currentState);
    const event = ctx.nextEvent();
    if (event[0]) {
      const eventNode = event[1];
      if (state.eventActLookup.has(eventNode.event)) {
        const actionNode = state.eventActLookup.get(eventNode.event);
        switch (actionNode.eventAction) {
          case EventAction.Goto:
            if (this.stateLookup.has(actionNode.state)) {
              state.state.exit(eventNode.event, ctx, eventNode.param);
              ctx.currentState = actionNode.state;
              const stateNext = this.stateLookup.get(ctx.currentState);
              stateNext.state.entry(
                eventNode.event,
                ctx,
                eventNode.param);
            } else {

            }
            ctx.accept();
            this.handleEvent(ctx);
            break;

          case EventAction.Do:
            actionNode.action.execute(
              eventNode.event,
              ctx,
              eventNode.param);
            ctx.accept();
            this.handleEvent(ctx);
            break;

          case EventAction.Defer:
            ctx.skipNext();
            this.handleEvent(ctx);
            break;

          case EventAction.Ignore:
            ctx.accept();
            this.handleEvent(ctx);
            break;
        }
      } else {
        ctx.accept();
        console.log(`Unable to find event for ${eventNode.event}`);
      }
    }
  }

}

class StateNode<TState, TEvent, TCtx extends BaseContext<TState, TEvent, TParam>, TParam> {

  eventActLookup = new Map<TEvent, EventNode<TState, TEvent, TCtx, TParam>>();

  constructor(
    public state: IState<TState, TEvent, TCtx, TParam>
  ) {
    this.state.events().forEach(eventNode => {
      if (!this.eventActLookup.has(eventNode.event)) {
        this.eventActLookup.set(eventNode.event, eventNode);
      } else {
        throw Error(`Already registered event ${eventNode.event}`);
      }
    });
  }
}

File: Mrh.Concurrent/StateMachine/StateMachine.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using NLog;

namespace Mrh.Concurrent.StateMachine
{
    public class StateMachine<TState, TEvent, TCtx, TParam> 
        where TState: struct 
        where TEvent: struct
        where TCtx: AbstractStateCtx<TState, TEvent, TParam>
    {

        private static readonly ILogger log = LogManager.GetCurrentClassLogger();
        
        private readonly Dictionary<TState, StateNode> states = new Dictionary<TState, StateNode>(10);

        /// <summary>
        ///     Used to register the context with the state machine.
        /// </summary>
        /// <param name="ctx">The context to register.</param>
        public void RegisterCtx(
            TCtx ctx)
        {
            ctx.NewEvent.Subscribe((x) =>
            {
                Task.Run(async () =>
                {
                    await this.HandleNextEvent(ctx);
                });
            });
        }

        /// <summary>
        ///     Triggers an event on the state machine.
        /// </summary>
        /// <param name="ctx">The state machine context.</param>
        /// <param name="event">The event to fire.</param>
        /// <param name="param">The parameter to the state machine.</param>
        /// <returns></returns>
        public bool Transition(
            TCtx ctx,
            TEvent @event,
            TParam param)
        {
            return ctx.Add(@event, param);
        }

        /// <summary>
        ///     Used to add a state to the state machine.
        /// </summary>
        /// <param name="state">The state to be registered. </param>
        /// <returns>The state machine.</returns>
        public StateMachine<TState, TEvent, TCtx, TParam> Add(
            IState<TState, TEvent, TCtx, TParam> state)
        {
            if (this.states.ContainsKey(state.State))
            {
                throw new StateAlreadyRegisteredException<TState>(state.State);
            }

            this.states[state.State] = new StateNode(state);
            return this;
        }

        private async Task HandleNextEvent(TCtx ctx)
        {
            if (ctx.Next(out var @event))
            {
                if (!this.states.TryGetValue(ctx.CurrentState, out var stateNode))
                {
                    throw new UnableToFindStateException<TState>(ctx.CurrentState);
                }
                if (stateNode.TryGetNode(@event.Event, out var eventAction))
                {
                    switch (eventAction.EventAction)
                    {
                        case EventActionType.Defer:
                            throw new ArgumentException("Deferred currently not supported.");
                        
                        case EventActionType.Do:
                            await eventAction.Action.Execute(
                                @event.Event,
                                ctx,
                                @event.Param);
                            break;
                        
                        case EventActionType.Goto:
                            await stateNode.State.Exit(
                                @event.Event,
                                ctx,
                                @event.Param);
                            ctx.CurrentState = eventAction.State.Value;
                            if (this.states.TryGetValue(ctx.CurrentState, out var node))
                            {
                                await node.State.Entry(
                                    @event.Event,
                                    ctx,
                                    @event.Param);
                            }
                            break;
                        
                        case EventActionType.Ignore:
                            break;
                    }
                }
            }
        }

        private class StateNode
        {
            private readonly Dictionary<TEvent, EventNode<TState, TEvent, TCtx, TParam>> eventActLookup = new Dictionary<TEvent, EventNode<TState, TEvent, TCtx, TParam>>(10);

            public readonly IState<TState, TEvent, TCtx, TParam> State;

            public StateNode(
                IState<TState, TEvent, TCtx, TParam> state)
            {
                this.State = state;
                foreach (var @event in state.Events)
                {
                    if (this.eventActLookup.ContainsKey(@event.Event))
                    {
                        throw new Exception($"Duplicate event registered {@event.Event}.");
                    }
                    this.eventActLookup[@event.Event] = @event;
                }
            }

            public bool TryGetNode(
                TEvent @event,
                out EventNode<TState, TEvent, TCtx, TParam> eventNode)
            {
                return this.eventActLookup.TryGetValue(
                    @event,
                    out eventNode);
            }
        }
    }
}